File: src/py_srvcli/py_srvcli/service_practice_b.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose
from turtlesim.srv import SetPen 

class ServicePracticeB(Node):

    def __init__(self):
        super().__init__('service_practice_b')
        self.velocity_publisher = self.create_publisher(Twist, 'turtle1/cmd_vel', 10)
        self.pose_subscriber = self.create_subscription(Pose, 'turtle1/pose', self.pose_callback, 10)
        self.pen_service_client = self.create_client(SetPen, 'turtle1/set_pen')  # Cambié a SetPen
        self.pose = Pose()
        self.is_near_wall = False
        self.color = 'green'

    def pose_callback(self, msg):
        self.pose = msg
        self.check_boundaries()
        self.update_pen_color()
        self.move_turtle()

    def check_boundaries(self):
        if self.pose.x < 1.0 or self.pose.x > 10.0 or self.pose.y < 1.0 or self.pose.y > 10.0:
            self.is_near_wall = True
        else:
            self.is_near_wall = False

    def update_pen_color(self):
        if self.pose.x > 5.5 and self.color != 'red':
            self.set_pen_color(255, 0, 0)  # Rojo
            self.color = 'red'
            self.get_logger().info('Set pen color to red')
        elif self.pose.x <= 5.5 and self.color != 'green':
            self.set_pen_color(0, 255, 0)  # Verde
            self.color = 'green'
            self.get_logger().info('Set pen color to green')

    def set_pen_color(self, r, g, b):
        if self.pen_service_client.wait_for_service(timeout_sec=1.0):
            req = SetPen.Request()
            req.r = r
            req.g = g
            req.b = b
            req.width = 3  # Ancho del bolígrafo
            req.off = 0  # Boligrafo encendido
            self.pen_service_client.call_async(req)
        else:
            self.get_logger().error('Service not available')

    def move_turtle(self):
        vel_msg = Twist()
        vel_msg.linear.x = 0.9 if self.is_near_wall else 1.0
        vel_msg.angular.z = 0.9 if self.is_near_wall else 0.0
        self.velocity_publisher.publish(vel_msg)

def main(args=None):
    rclpy.init(args=args)
    service_practice_b_node = ServicePracticeB()
    rclpy.spin(service_practice_b_node)
    service_practice_b_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/cpp_srvcli/src/service_practice_b.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "turtlesim/msg/pose.hpp"
#include "turtlesim/srv/set_pen.hpp"

class ServicePracticeB : public rclcpp::Node {
public:
    ServicePracticeB() : Node("service_practice_b"), is_near_wall(false), color("green") {
        velocity_publisher = this->create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", 10);
        pose_subscriber = this->create_subscription<turtlesim::msg::Pose>(
            "turtle1/pose", 10, std::bind(&ServicePracticeB::pose_callback, this, std::placeholders::_1)
        );
        pen_service_client = this->create_client<turtlesim::srv::SetPen>("turtle1/set_pen");
    }

private:
    void pose_callback(const turtlesim::msg::Pose::SharedPtr msg) {
        pose = *msg;
        check_boundaries();
        update_pen_color();
        move_turtle();
    }

    void check_boundaries() {
        if (pose.x < 1.0 || pose.x > 10.0 || pose.y < 1.0 || pose.y > 10.0) {
            is_near_wall = true;
        } else {
            is_near_wall = false;
        }
    }

    void update_pen_color() {
        if (pose.x > 5.5 && color != "red") {
            set_pen_color(255, 0, 0);  // Rojo
            color = "red";
            RCLCPP_INFO(this->get_logger(), "Set pen color to red");
        } else if (pose.x <= 5.5 && color != "green") {
            set_pen_color(0, 255, 0);  // Verde
            color = "green";
            RCLCPP_INFO(this->get_logger(), "Set pen color to green");
        }
    }

    void set_pen_color(uint8_t r, uint8_t g, uint8_t b) {
        if (pen_service_client->wait_for_service(std::chrono::seconds(1))) {
            auto req = std::make_shared<turtlesim::srv::SetPen::Request>();
            req->r = r;
            req->g = g;
            req->b = b;
            req->width = 3;  // Ancho del bolígrafo
            req->off = 0;  // Bolígrafo encendido
            pen_service_client->async_send_request(req);
        } else {
            RCLCPP_ERROR(this->get_logger(), "Service not available");
        }
    }

    void move_turtle() {
        auto vel_msg = geometry_msgs::msg::Twist();
        vel_msg.linear.x = is_near_wall ? 0.9 : 1.0;
        vel_msg.angular.z = is_near_wall ? 0.9 : 0.0;
        velocity_publisher->publish(vel_msg);
    }

    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr velocity_publisher;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose_subscriber;
    rclcpp::Client<turtlesim::srv::SetPen>::SharedPtr pen_service_client;
    turtlesim::msg::Pose pose;
    bool is_near_wall;
    std::string color;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ServicePracticeB>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}